sql/pkg/repository/memory/memory.go

```diff
@@ -0,0 +1,162 @@
+
+	return contact, nil
+}
+
+func (r *ContactsRepositoryInMemory) ListAll() ([]model.Contact, error) {
+	var result []model.Contact
+	row, err := r.storage.Query("select (ID, FirstName, LastName, Phone, Email) from Contacts")
+	if err != nil {
+		return result, err
+	}
+
+	defer row.Close()
+	for row.Next() {
+		var tmp model.Contact
+		err := row.Scan(&tmp.ID, &tmp.FirstName, &tmp.LastName, &tmp.Phone, &tmp.Email)
+		if err != nil {
+			return result, err
+		}
+
+		result = append(result, tmp)
+	}
+
+	return result, nil
+}
+
+func (r *ContactsRepositoryInMemory) GetByID(id uint) (model.Contact, error) {
+	var contact model.Contact
+	row, err := r.storage.Query("select (ID, FirstName, LastName, Phone, Email) from Contacts where ID = ?", id)
+	if err != nil {
+		return model.Contact{}, nil
+	}
+
+	defer row.Close()
+	err = row.Scan(&contact.ID, &contact.FirstName, &contact.LastName, &contact.Phone, &contact.Email)
+	if err != nil {
+		return contact, nil
+	}
+
+	return contact, nil
+}
+
+func (r *ContactsRepositoryInMemory) GetByPhone(phone string) (model.Contact, error) {
+	var contact model.Contact
+	row, err := r.storage.Query("select (ID, FirstName, LastName, Phone, Email) from Contacts where ID = ?", phone)
+	if err != nil {
+		return model.Contact{}, nil
+	}
+
+	defer row.Close()
+	err = row.Scan(&contact.ID, &contact.FirstName, &contact.LastName, &contact.Phone, &contact.Email)
+	if err != nil {
+		return contact, nil
+	}
+
+	return contact, nil
+}
+
+func (r *ContactsRepositoryInMemory) GetByEmail(email string) (model.Contact, error) {
+	var contact model.Contact
+	row, err := r.storage.Query("select (ID, FirstName, LastName, Phone, Email) from Contact where Email = ?", email)
+	if err != nil {
+		return contact, err
+	}
+
+	defer row.Close()
+	err = row.Scan(&contact.ID, &contact.FirstName, &contact.LastName, &contact.Phone, &contact.Email)
+	if err != nil {
+		return contact, nil
+	}
+
+	return model.Contact{}, fmt.Errorf("record not found")
+}
+
+func (r *ContactsRepositoryInMemory) SearchByName(name string) ([]model.Contact, error) {
+	var contacts []model.Contact
+	row, err := r.storage.Query("select (ID, FirstName, LastName, Phone, Email) from Contacts where Email = ?", name)
+	if err != nil {
+		return contacts, nil
+	}
+
+	defer row.Close()
+	for row.Next() {
+		var tmp model.Contact
+		err := row.Scan(&tmp.ID, &tmp.FirstName, &tmp.LastName, &tmp.Phone, &tmp.Email)
+		if err != nil {
+			return contacts, err
+		}
+
+		contacts = append(contacts, tmp)
+	}
+	return contacts, nil
+}
+
+func (r *ContactsRepositoryInMemory) Delete(id uint) error {
+	info, err := r.storage.Exec("delete from Contacts where id = ?", id)
+	if err != nil {
+		return err
+	}
+
+	count, _ := info.RowsAffected()
+
+	if count > 1 {
+		return errors.New("deleted " + string(count) + " strings, but expected one")
```

Comment:
It's not needed to return an error in case of 0 affected rows, the most common option to implement delete, return nil error if entity have already don't exist